pkg/compactor/deleted_series_cleaner.go

```diff
@@ -0,0 +1,462 @@
+	// If after iterating through all blocks, the value is 0
+	// Then the deletion request can be updated to processed state
+	blocksLeftPerRequest := make(map[string]int, len(pendingDeletion))
+
+	var pendingBlocks int = 0
+
+	err = userBucket.Iter(ctx, "", func(name string) error {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		id, ok := block.IsBlockDir(name)
+		if !ok {
+			return nil
+		}
+
+		// If the block was marked for deletion, then skip it
+		deletionMarkFile := path.Join(id.String(), metadata.DeletionMarkFilename)
+		deletionMarkExists, err := userBucket.Exists(ctx, deletionMarkFile)
+		if err != nil {
+			c.blocksFailedTotal.Inc()
+			level.Error(userLogger).Log("msg", "failed to check if deletion mark exists for block", "block", id, "err", err)
+			// Since we don't know if this block has been deleted or not, we should not proceed with series deletion.
+			// To be safe, increase the number of blocks per request so that the tombstones aren't moved to processed
+			// state while there still remains a block for deletion.
+			for _, t := range pendingDeletion {
+				blocksLeftPerRequest[t.RequestID] += 1
+			}
+			return nil //continue with the other blocks
+		}
+		if deletionMarkExists {
+			return nil
+		}
+
+		meta, err := block.DownloadMeta(ctx, userLogger, userBucket, id)
+		if err != nil {
+			c.blocksFailedTotal.Inc()
+			level.Error(userLogger).Log("msg", "failed to download block meta.json file", "block", id, "err", err)
+			// don't know if this block would have required deletion, so to be safe, need to increase the
+			// blocks left per request. Otherwise, the cleaner might think that all blocks have been re-written and
+			// the request can be moved to processed when there is still a block that requires deletion.
+			for _, t := range pendingDeletion {
+				blocksLeftPerRequest[t.RequestID] += 1
+			}
+			return nil //continue with the other blocks
+		}
+
+		// Keep track of which deletions have already been applied to this block
+		deletionRequestsApplied := make(map[string]struct{})
+		if meta.Thanos.Rewrites != nil {
+			for _, r := range meta.Thanos.Rewrites {
+				for _, d := range r.DeletionsApplied {
+					deletionRequestsApplied[d.RequestID] = struct{}{}
+				}
+			}
+		}
+
+		blockForDeletion := false
+		var deletions []metadata.DeletionRequest
+		for _, t := range pendingDeletion {
+			if meta.MinTime > t.EndTime || t.StartTime > meta.MaxTime {
+				// tombstone time doesn't overlap with the block timestamps
+				continue
+			}
+
+			if _, processed := deletionRequestsApplied[t.RequestID]; processed {
+				// Request has already been processed for this block
+				continue
+			}
+
+			blockForDeletion = true
+			blocksLeftPerRequest[t.RequestID] += 1
+
+			if time.Since(t.GetCreateTime()) < c.purgerCfg.DeleteRequestCancelPeriod {
+				// the cancellation period is not over yet
+				continue
+			}
+
+			if meta.Compaction.Level < len(c.cfg.CompactionBlockRanges) {
```

Comment:
Question: Compation.Level is 1,2,3 etc right? Isn't CompactionBlockRange 2h, 12h, 24h etc?

Reply:
Yes that is correct. The compaction time range in the block `maxTime-minTime` isn't always going to be exactly 2h,12h,24h depending on the samples in the block. Which is why I used the compaction level instead. 

The compaction level is always increased when the block is compacted. It starts at 1 for the 2h block, then increased to level 2 for the 6 hour block, and then to 3 for the 24 hour block.